File: src/main/java/repositories/CategoryRepository.java
package repositories;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import model.Category;
import util.URLHelper;

import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;

public class CategoryRepository {

    private List<Category> categories;

    public CategoryRepository() {
    }

    public void downloadCategories() {
        try (InputStream inputStream = URLHelper.getData("https://opentdb.com/api_category.php", "GET")){
            ObjectMapper objectMapper = new ObjectMapper();
            HashMap<String, List<Category>> allCategories = objectMapper.readValue(inputStream, new TypeReference<>() {});
            this.categories = allCategories.get("trivia_categories");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public List<Category> getCategories() {
        return categories;
    }

    public void setCategories(List<Category> categories) {
        this.categories = categories;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CategoryRepository that = (CategoryRepository) o;
        return Objects.equals(categories, that.categories);
    }

    @Override
    public int hashCode() {
        return Objects.hash(categories);
    }

    @Override
    public String toString() {
        return "CategoryRepository{" +
                "categories=" + categories +
                '}';
    }
}

File: src/main/webapp/scripts/repositories/categoryRepository.js
import { Category } from '../model/category.js';
import fetch from 'node-fetch';

class CategoryRepository {
    constructor() {
        this.categories = [];
    }

    async downloadCategories() {
        const url = `https://opentdb.com/api_category.php`;
        try {
            const response = await fetch(url);
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            const { trivia_categories } = await response.json();
            this.categories = trivia_categories.map(category => new Category(category.id, category.name));
            return this.categories;
        } catch (error) {
            throw new Error(`Failed to download categories: ${error}`);
        }
    }

    getCategories() {
        return this.categories;
    }

    setCategories(categories) {
        this.categories = categories;
    }

    toString() {
        return `CategoryRepository{categories=${this.categories}}`;
    }
}

export { CategoryRepository };
